rust/agama-manager/src/hardware.rs

```diff
@@ -0,0 +1,288 @@
+    }
+
+    /// Converts the information to a HardwareInfo struct.
+    pub fn to_hardware_info(&self) -> HardwareInfo {
+        let Some(root) = &self.root else {
+            return HardwareInfo::default();
+        };
+
+        HardwareInfo::from(root)
+    }
+}
+
+/// Hardware information from the underlying system.
+///
+/// It relies on lshw to read the hardware information.
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, PartialEq)]
+struct HardwareNode {
+    pub id: String,
+    pub class: String,
+    pub claimed: Option<bool>,
+    pub description: Option<String>,
+    pub vendor: Option<String>,
+    pub product: Option<String>,
+    pub version: Option<String>,
+    pub serial: Option<String>,
+    pub businfo: Option<String>,
+    pub dev: Option<String>,
+    pub driver: Option<String>,
+    pub physid: Option<String>,
+    pub size: Option<u64>,
+    pub capacity: Option<u64>,
+    #[serde(default)]
+    #[serde_as(as = "OneOrMany<_, PreferMany>")]
+    pub logicalname: Vec<String>,
+    pub configuration: Option<serde_json::Value>,
+    #[serde(default)]
+    pub capabilities: Option<serde_json::Value>,
+    #[serde(default)]
+    pub children: Vec<HardwareNode>,
+}
+
+impl HardwareNode {
+    /// Searches hardware information using the id (e.g., "cpu").
+    ///
+    /// It assumes that the id is unique.
+    ///
+    /// * `id`: id to search for (e.g., "cpu", "memory", etc.).
+    pub fn find_by_id(&self, id: &str) -> Option<&HardwareNode> {
+        if self.id == id {
+            return Some(&self);
+        }
+
+        for children in &self.children {
+            let result = children.find_by_id(id);
+            if result.is_some() {
+                return result;
+            }
+        }
+
+        None
+    }
+
+    /// Searches hardware information by class (e.g., "disk").
+    ///
+    /// It might be multiple elements of the same class.
+    ///
+    /// * `class`: class to search for (e.g., "disk", "processor", etc.).
+    pub fn find_by_class(&self, class: &str) -> Vec<&HardwareNode> {
+        let mut results = vec![];
+        self.search_by_class(class, &mut results);
+        results
+    }
+
+    fn search_by_class<'a>(&'a self, class: &str, results: &mut Vec<&'a HardwareNode>) {
+        if self.class == class {
+            results.push(&self);
+        }
+
+        for children in &self.children {
+            children.search_by_class(class, results);
+        }
+    }
+}
+
+impl From<&HardwareNode> for HardwareInfo {
+    fn from(value: &HardwareNode) -> Self {
+        // let system_info = value.find_by_class("system");
+        let cpu = value
+            .find_by_class("processor")
+            .first()
+            .and_then(|c| c.product.clone());
+
+        let memory = value.find_by_id("memory").and_then(|m| m.size);
+
+        let model = value.find_by_class("system").first().map(|s| {
+            format!(
+                "{} {}",
+                s.vendor.clone().unwrap_or_default(),
```

Comment:
I am not sure if it is better to use empty string like here or explicit Unknown key so it is obvious that we do not know verdor or version.